syntax = "proto3";

package brax;

// A point or scalar value in 3d space.
message Vector3 {
  float x = 1;
  float y = 2;
  float z = 3;
}

// Prevents motion or rotation along specifed axes.
message Frozen {
  // Freeze motion along the x, y, or z axes.
  Vector3 position = 1;
  // Freeze rotation around the x, y, or z axes.
  Vector3 rotation = 2;
  // Override all the position and rotation fields, setting them to 1.
  bool all = 3;
}

// Bodies have a rigid shape, mass, and rotational inertia. Bodies may connect
// to other bodies via joints, forming a kinematic tree.
message Body {
  // Unique name for this body
  string name = 1;
  // Geometric primitives that define the shape of the body
  repeated Collider colliders = 2;
  // Diagonal of the inertia tensor of this body about its center-of-mass
  Vector3 inertia = 3;
  // Mass of this body in kg
  float mass = 4;
  // Prevents motion or rotation along specified axes for this body.
  Frozen frozen = 5;
}

// Primitive shape that composes the collision surface of a body.
message Collider {
  // Material properties that influence collision behavior
  message Material {
    // Bounciness of this material
    float elasticity = 1;
    // How much this material resists translation when in contact
    float friction = 2;
  }
  // A 6-sided rectangular prism
  message Box {
    // Half the size of the box in each dimension.
    Vector3 halfsize = 1;
  }
  // An infinite plane with normal vector (0, 0, 1)
  message Plane {}
  // A clipped plane.
  message ClippedPlane {
    float halfsize_x = 1;
    float halfsize_y = 2;
  }
  // A sphere
  message Sphere {
    float radius = 1;
  }
  // A cylinder with rounded ends.
  message Capsule {
    // Radius of the sphere at each rounded end
    float radius = 1;
    // End-to-end length of the capsule
    float length = 2;
    // Capsule end (0: both ends, 1: top end, -1: bottom end)
    int32 end = 3;
  }
  // A height map aligned with the x-y plane
  message HeightMap {
    // The width and length of the square height map.
    float size = 1;
    // A flattened square matrix of the height data in row major order where
    // the row index is the x-position and the column index is the y-position.
    repeated float data = 2;
  }
  // Physical material of the collider
  Material material = 8;
  // A mesh. Currently, only mesh-plane and mesh-capsule collisions are
  // supported.
  message Mesh {
    // Name of the mesh geometry defined in the config.
    string name = 1;
    // Scaling for the mesh.
    float scale = 2;
  }
  // Position relative to parent body
  Vector3 position = 1;
  // Rotation relative to parent body
  Vector3 rotation = 2;
  // A collider may only be one type
  oneof type {
    Box box = 3;
    Plane plane = 4;
    Sphere sphere = 5;
    Capsule capsule = 6;
    HeightMap heightMap = 7;
    Mesh mesh = 9;
    ClippedPlane clipped_plane = 13;
  }
  // Color of the collider in css notation (e.g. '#ff0000' or 'red')
  string color = 10;
  // A hidden collider is not visualized
  bool hidden = 11;
  // A no-contact collider doesn't yield any collisions
  bool no_contact = 12;
}

// Joints connect two bodies and restrain their movement to a rotational frame.
message Joint {
  // Min/max tuple.
  message Range {
    float min = 1;
    float max = 2;
  }
  // Unique name for this joint
  string name = 1;
  // Spring constant controlling the strength of this joint
  // If not set, defaults to infinite strength.
  float stiffness = 2;
  // Parent body's name
  string parent = 3;
  // Child body's name
  string child = 4;
  // Offset from the parent's center of mass to the joint
  Vector3 parent_offset = 5;
  // Offset from the child's center of mass to the joint
  Vector3 child_offset = 6;
  // The offset rotation from the parent to the joint's coordinate system
  // The joint's local x, y, and z axes are made active by
  // supplying 1, 2, or 3 angle_limits, respectively.
  Vector3 rotation = 7;
  // Damps parent and child angular velocities to be equal
  float angular_damping = 8;
  // Limits for actuated axes.  Supply one Range for joint-local-x actuation
  // and two for x,y.
  repeated Range angle_limit = 9;
  // Strength of the force that restores joint to within the angle limits
  // If stiffness is not set, this field is ignored.
  // If stiffness is non-zero, and this field is not set, defaults to a
  // joint-dependent fraction of `stiffness`.
  optional float limit_strength = 10;
  // Damps the spring connecting the parent to the child
  // If stiffness is not set, this field is ignored.
  // If stiffness is non-zero, and this field is not set, defaults to a
  // joint-dependent fraction of `sqrt(stiffness)`.
  optional float spring_damping = 11;
  // The initial rotation between the parent and child, defined in terms of
  // an intrinsic euler rotation about parent's local axes.  This sets the
  // reference "zero" of the joint's rotation axes.
  Vector3 reference_rotation = 12;
}

// Actuators provide interactive control of a system via the step function.
// An actuator acts on a single joint.  It occupies as many dimensions in
// the action vector as there are ranges in its joint.
message Actuator {
  // Unique name for this actuator
  string name = 1;
  // Joint for this actuator
  string joint = 2;
  // Spring constant multiplier on input value to output torque
  float strength = 3;
  // Applies a direct torque to the joint
  message Torque {}
  // Drives a joint to a desired target angle of rotation about its axes
  message Angle {}
  // An actuator may only be one type
  oneof type {
    Torque torque = 4;
    Angle angle = 5;
  }
}

// Forces provide interactive control of a system via the step function.
// A force acts on a single body.  It occupies 3 positions in the action vector
// corresponding to the direction and magnitude of the force to apply, which is
// additionally multiplied by `strength`.
message Force {
  // Unique name for this force
  string name = 1;
  // Body for this force
  string body = 2;
  // Multiplier for intensity of force
  float strength = 3;
  // Applies a force to the center of the body in the specified direction
  message Thruster {}
  // Applies a torque to the center of the body around a specified axis
  message Twister {}
  // A force may only be one type
  oneof type {
    Thruster thruster = 4;
    Twister twister = 5;
  }
}

// DefaultState determines the default orientations of bodies in the system
// returned by the default_qp() function.
message DefaultState {
  message JointAngle {
    // Joint name
    string name = 1;
    // Default angles in degrees
    Vector3 angle = 2;
  }
  message QP {
    // Name of body
    string name = 1;
    // Position of body in world space
    Vector3 pos = 2;
    // Rotation of body in world space
    Vector3 rot = 3;
    // Velocity of body in world space
    Vector3 vel = 4;
    // Angular velocity of body in world space
    Vector3 ang = 5;
  }
  // Default joint angles.
  repeated JointAngle angles = 1;
  // Default orientations for root bodies (not a child of a joint).
  repeated QP qps = 2;
}

// Geometry of a mesh.
message MeshGeometry {
  // Name of the mesh geometry. This is used in mesh colliders to refer to the
  // geometry.
  string name = 1;
  // Path of the mesh file. See https://trimsh.org/ for the supported formats.
  // If the path is specified, then the {vertices, faces, vertex_normals}
  // fields below will be ignored and populated from the mesh defined in the
  // file.
  string path = 2;
  // Vertices of the mesh.
  repeated Vector3 vertices = 3;
  // Triangular faces. This will be a flattened array of triples that contain
  // the indices of the vertices in the `vertices` field above. For example,
  //
  // vertices { x: -0.5 y: -0.5 z: 0 }
  // vertices { x: +0.5 y: -0.5 z: 0 }
  // vertices { x: +0.5 y: +0.5 z: 0 }
  // vertices { x: -0.5 y: +0.5 z: 0 }
  // vertices { x: 0 y: 0 z: 1.0 }
  // faces: [0, 2, 1, 0, 3, 2, 0, 4, 3, 0, 1, 4, 1, 2, 4, 2, 3, 4]
  //
  // defines a pyramid with 6 faces (two for the bottom and four for the
  // sides).
  repeated int32 faces = 4;
  // Unit normal vectors for each vertex.
  repeated Vector3 vertex_normals = 5;
  // Unit normal vectors for each face.
  repeated Vector3 face_normals = 6;
}

// The configuration of a system for Brax to simulate
// Next ID: 22
message Config {
  message NamePair {
    string first = 1;
    string second = 2;
  }
  // All of the bodies in the system
  repeated Body bodies = 1;
  // All of the joints in the system
  repeated Joint joints = 2;
  // All of the actuators in the system
  repeated Actuator actuators = 3;
  // Forces to apply to bodies
  repeated Force forces = 16;
  // Default bounciness of collisions, for colliders that don't specify
  // elasticity.
  float elasticity = 4;
  // Default friction (how much surfaces in contact resist translation),
  // for colliders that don't specify friction.
  float friction = 5;
  // Uniform force applied to each body
  Vector3 gravity = 6;
  // Linear velocity damping applied to each body
  float velocity_damping = 7;
  // Angular velocity damping applied to each body
  float angular_damping = 8;
  // How aggressively interpenetrating bodies should push away from one another
  float baumgarte_erp = 9;
  // Pairs of bodies in the scene than can collide with one another
  repeated NamePair collide_include = 10;
  // Amount of time to simulate each step, in seconds
  float dt = 11;
  // How many substeps to perform to maintain numerical stability
  int32 substeps = 12;
  // Prevents motion or rotation along specified axes for the entire system
  Frozen frozen = 13;
  // Default states of bodies in the system
  repeated DefaultState defaults = 14;
  // Limits the number of collision checks per collider pair
  int32 collider_cutoff = 15;
  // All of the mesh geometries in the system.
  repeated MeshGeometry mesh_geometries = 17;
  // The physics used for this config.  Options: ['pbd', 'legacy_spring']
  // Defaults to 'pbd'
  string dynamics_mode = 18;
  // Sets the default magnitude of position based updates for position data
  optional float solver_scale_pos = 19;
  // Sets the default magnitude of position based updates for angular data
  optional float solver_scale_ang = 20;
  // Sets the default magnitude of position based updates for collision data
  optional float solver_scale_collide = 21;
}
